### PR TITLE
Add eager bool functions: Bool.(logand, logor, logxor).

### PR DESCRIPTION
**Pull request by yallop**
### Motivation

This PR is based on a request from Oleg Kiselyov, who mentioned recently that it would be useful to have eager logical boolean operations as well as the short-circuiting (lazy) versions.  The lazy versions generate code with conditional jumps that might sometimes be slower than a sequence of bitwise operations.

### Implementation

Branchless implementations of the operators are easy, since we can just reuse the existing primitives for int:

```ocaml
external logand : bool -> bool -> bool = "%andint"
external logor : bool -> bool -> bool = "%orint"
external logxor : bool -> bool -> bool = "%xorint"
```

### Benchmarking

The eager operators are sometimes faster than the lazy operators.  [Here's](https://gist.github.com/yallop/ee2c75a54204ba6670598b33d15ba310) a benchmark that tests eager and lazy versions of [`Uchar.is_valid`](https://github.com/ocaml/ocaml/blob/d90601c59028f8045bb26d37f546070f5a212549/stdlib/uchar.ml#L43) on an array of integers, half of which are valid.  On my system, with OCaml 5.2.0, the results are as follows:

| Name       | Time/run | Percentage |
|------------|----------|------------|
| loop_lazy  | 67.46us  | 100.00%    |
| loop_eager | 41.88us  | 62.08%     |

### Other implementations

There's an implementation in Base ([here](https://github.com/janestreet/base/blob/de1f1123/src/bool.ml#L75-L81)) and an unoptimized implementation in Batteries with funny names (`logand` is called [`mul`](https://github.com/ocaml-batteries-team/batteries-included/blob/00c591ea/src/batBool.ml#L45) and `logor` is called [`add`](https://github.com/ocaml-batteries-team/batteries-included/blob/00c591ea/src/batBool.ml#L44)).

### CODE REVIEW

File: stdlib/bool.mli
Comment (nojb) [inline stdlib/bool.mli]: Just a suggestion but I would prefer a more "concrete" docstring:
```suggestion
(** [logand b1 b2] is [true] if and only if [b1] and [b2] are both [true]. *)
```
In particular, I think using "e1", "e2", "expression", ... is a bit misleading here; I would prefer "b1", "b2", "boolean", ... instead.

File: Changes
Comment (nojb) [inline Changes]:
```suggestion
- ?: Add eager boolean operations Bool.logand, Bool.logor, Bool.logxor
```

**Review by nojb — APPROVED** [review]
LGTM, modulo suggestion below.

A second approval is needed before the PR can be merged.

**Review by lthls — COMMENTED** [review]
I tried with 5.2.0, forcing inlining of the `is_valid` functions in the benchmark, and I get improbable results (the eager version is 15 to 20 times faster). It turns out that the whole loop body is removed; I suspect a bug somewhere in the native backend (the backend probably makes some reasonable assumptions that this PR breaks).
Please don't merge until I understand what's going on. Otherwise, I'm in favour of the PR.

**Comment by lthls** [comment]
> I tried with 5.2.0, forcing inlining of the `is_valid` functions in the benchmark, and I get improbable results (the eager version is 15 to 20 times faster). It turns out that the whole loop body is removed; I suspect a bug somewhere in the native backend (the backend probably makes some reasonable assumptions that this PR breaks).

Sorry for the noise; it turns out that the eager version can figure out that the counter isn't used and doesn't compute it at all. I suppose the branches in the lazy version prevent this optimisation from applying there.
With a well-placed `Sys.opaque_identity`, I get reasonable numbers again.

**Review by lthls — APPROVED** [review]

**Review by dra27 — APPROVED** [review]
Is it worth being explicit that `logand` evaluates `b2` even if `b1` evaluates to `false` and likewise for `logor`?

**Comment by nojb** [comment]
> Is it worth being explicit that `logand` evaluates `b2` even if `b1` evaluates to `false` and likewise for `logor`?

My feeling is that the extra explanation may be more confusing than clarifying to the casual reader, as that is the normal behaviour of functions; so personally I would leave it as it is.

Except this minor point, the PR now has three official approvals, hence it is good to go!

**Pull request merged**
